src/Microsoft.OData.Core/Json/JsonValueUtilsAsync.cs

```diff
@@ -0,0 +1,491 @@
+        /// <param name="stringEscapeOption">The string escape option.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        /// <param name="arrayPool">Array pool for renting a buffer.</param>
+        internal static async Task WriteValueAsync(this TextWriter writer, string value, ODataStringEscapeOption stringEscapeOption, Ref<char[]> buffer, ICharArrayPool arrayPool = null)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            if (value == null)
+            {
+                await writer.WriteAsync(JsonConstants.JsonNullLiteral).ConfigureAwait(false);
+            }
+            else
+            {
+                await writer.WriteEscapedJsonStringAsync(value, stringEscapeOption, buffer, arrayPool).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Write a byte array.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">Byte array to be written.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        /// <param name="arrayPool">Array pool for renting a buffer.</param>
+        internal static async Task WriteValueAsync(this TextWriter writer, byte[] value, Ref<char[]> buffer, ICharArrayPool arrayPool = null)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            if (value == null)
+            {
+                await writer.WriteAsync(JsonConstants.JsonNullLiteral).ConfigureAwait(false);
+            }
+            else
+            {
+                await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);
+                await writer.WriteBinaryStringAsync(value, buffer, arrayPool).ConfigureAwait(false);
+                await writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Write a byte array.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">Byte array to be written.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        /// <param name="arrayPool">Array pool for renting a buffer.</param>
+        internal static async Task WriteBinaryStringAsync(this TextWriter writer, byte[] value, Ref<char[]> buffer, ICharArrayPool arrayPool)
+        {
+            Debug.Assert(writer != null, "writer != null");
+            Debug.Assert(value != null, "The value must not be null.");
+
+            buffer.Value = BufferUtils.InitializeBufferIfRequired(arrayPool, buffer.Value);
+            Debug.Assert(buffer.Value != null);
+
+            int bufferLength = buffer.Value.Length;
+
+            // Try to hold base64 string as much as possible in one converting.
+            int bufferByteSize = bufferLength * 3 / 4;

```

Comment:
Why ` 3/4`? maybe that should be a constant?

Reply:
@habbes Not a 100% sure. This was lifted from the synchronous workflow. I believe the idea behind using 3 quarters of the buffer is to leave an allowance.